esmvaltool/diag_scripts/land_carbon_cycle/diag_zonal_correlation.py

```diff
@@ -0,0 +1,452 @@
+    group_metadata,
+    run_diagnostic,
+)
+
+import esmvaltool.diag_scripts.land_carbon_cycle.extraUtils as xu
+from esmvaltool.diag_scripts.land_carbon_cycle.shared import (
+    _get_obs_data_zonal,
+    _load_variable,
+)
+from esmvaltool.diag_scripts.land_carbon_cycle.provenance import (
+    _get_ancestor_files,
+    get_provenance_record,
+)
+
+
+def _get_fig_config(diag_config):
+    '''
+    get the default settings of the figure, and replace default with
+    runtime settings from recipe
+
+    Arguments:
+        diag_config - nested dictionary of metadata
+
+    Returns:
+        a dot dictionary of settings
+    '''
+
+    fig_config = {
+        'fill_value': np.nan,
+        'correlation_method': 'pearson',
+        'min_points_frac': 0.125,
+        # define the data and information for plotting ratios
+        'ax_fs': 7.1,
+        'valrange_x': (-1, 1),
+        'valrange_y': (-70, 90),
+        'bandsize': 9.5,
+        'gpp_threshold': 0.01
+    }
+    fig_config.update(diag_config.get('fig_config'))
+    return fig_config
+
+
+def _apply_common_mask(dat_1, dat_2, dat_3):
+    '''
+    apply a common mask to three arrays so that they have the same locations of
+    all valid and invalid (non numeric) grid cells
+    '''
+    dat_1_mask = np.ma.getmask(np.ma.masked_invalid(dat_1))
+    dat_2_mask = np.ma.getmask(np.ma.masked_invalid(dat_2))
+    dat_3_mask = np.ma.getmask(np.ma.masked_invalid(dat_3))
+    _val_mask_a = 1 - (1 - dat_1_mask) * (1 - dat_2_mask) * (1 - dat_3_mask)
+    _val_mask = np.ma.nonzero(_val_mask_a)
+    dat_1[_val_mask] = np.nan
+    dat_2[_val_mask] = np.nan
+    dat_3[_val_mask] = np.nan
+    dat_1 = np.ma.masked_invalid(dat_1)
+    dat_2 = np.ma.masked_invalid(dat_2)
+    dat_3 = np.ma.masked_invalid(dat_3)
+    return dat_1, dat_2, dat_3
+
+
+def _partialCorr(dat_columns, fig_config):
+    '''
+    A function to calculate the linear partial correlation between the
+    variables in the first and second column of dat_columns controlled for the
+    covariation with that in the third column.
+
+    Arguments:
+        dat_columns - an array with different variables in different columns
+        fig_config - configuration with correlation_method
+
+    Returns:
+        r123 - correlation between variables 1 and 2 controlled for 3
+    '''
+    dat_x = dat_columns[:, 0]
+    dat_y = dat_columns[:, 1]
+    dat_z = dat_columns[:, 2]
+    if fig_config['correlation_method'] == 'pearson':
+        r12 = stats.pearsonr(dat_x, dat_y)[0]
+        r13 = stats.pearsonr(dat_x, dat_z)[0]
+        r23 = stats.pearsonr(dat_y, dat_z)[0]
+    elif fig_config['correlation_method'] == 'spearman':
+        r12 = stats.spearmanr(dat_x, dat_y)[0]
+        r13 = stats.spearmanr(dat_x, dat_z)[0]
+        r23 = stats.spearmanr(dat_y, dat_z)[0]
+    else:
+        sys.exit('set a valid correlation_method [pearson/spearman]')
+    r123 = (r12 - r13 * r23) / np.sqrt((1 - r13**2) * (1 - r23**2))
+    return r123
```

Comment:
also, it would be nice to document the controlled correlation formula

Reply:
OK.

Reply:
Done.